File: 250713/0300/minkyu.py
from bisect import bisect_left


class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        dp = [1] * len(nums)
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
    
    def lengthOfLIS2(self, nums: list[int]) -> int:
        lis = [nums[0]]
        for num in nums:
            idx = bisect_left(lis, num)
            if idx == len(lis) and num > lis[-1]:
                lis.append(num)
            else:
                lis[idx] = num
        return len(lis)

File: 250713/0300/minkyu.rs
impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let mut dp = vec![1; nums.len()];
        for i in 0..nums.len() {
            for j in 0..i {
                if nums[i] > nums[j] {
                    dp[i] = dp[i].max(dp[j] + 1);
                }
            }
        }
        *dp.iter().max().unwrap()
    }

    pub fn length_of_lis2(nums: Vec<i32>) -> i32 {
        let mut lis = vec![nums[0]];
        for num in nums {
            let res = lis.binary_search(&num);
            match res {
                Err(idx) => {
                    if idx == lis.len() {
                        lis.push(num);
                    } else {
                        lis[idx] = num;
                    }
                }
                _ => {}
            }
        }
        lis.len() as i32
    }
}
